Fix MapViz to allow the use of a non-mapbox style without raising a TokenError and add error propagation from JS

EtienneDesticourt opened the pull request:
Fixes #58.

akacarlyann commented:
Just a thought: I think you might need to include logic to account for a style url defined with prefix __https://api.mapbox.com__ (which gets transformed to __mapbox:__ in the templates).  Not sure if there are any other mapbox prefixes though.

EtienneDesticourt commented:
That's a good point.
Now that I think about it, might there be cases where the user would want to use a custom style but still rely on the mapbox API for other map functionalities? (I'm thinking below_layer, is that fetched online?)
In that case we might want to remove the check entirely and wait for it to fail at a lower level in the API once the token is actually used.

ryanbaumann commented:
@EtienneDesticourt Thanks for the PR!  I'd like to add one more requirement to merge this PR, specifically https://github.com/mapbox/mapboxgl-jupyter/issues/19.  Is that something you could hit?  

> Now that I think about it, might there be cases where the user would want to use a custom style but still rely on the mapbox API for other map functionalities? (I'm thinking below_layer, is that fetched online?)

A user that specifies a custom style sheet could still use Mapbox resources, yes @EtienneDesticourt.  We can handle that with a proper error message, however.

`below_layer` does not require Mapbox API access - it's simply an identifier of what location in the base map layer stack (road labels, country labels, landcover, etc) to place the layer you're visualizing.

> In that case we might want to remove the check entirely and wait for it to fail at a lower level in the API once the token is actually used.

Yep, that's perfect here.

EtienneDesticourt commented:
@ryanbaumann 
Thanks for the clarifications.
Alright for #19, I'll try to take care of it over the weekend.

EtienneDesticourt commented:
@ryanbaumann 
I did the changes you specified in #19 and tried them out with a custom style url and custom style dict.  
I'm not sure how to write tests for them because I'm not familiar with Jinja and the way it deals with javascript exceptions.  

I also removed two tests that checked for raised TokenErrors since we don't raise them anymore when the token is missing.

Are you okay with the changes?

&nbsp;


On a side note, as a first time contributor to this project I thought it'd be useful if I took some notes on the problems I encountered, some of it is **really** minor, some of it not so much:  
1) For development we have to use "pip install mock" in addition to "pip install pytest", it's not specified in the README.
2) There is no specified style guide. I assumed pep 8 but it's good to have a style guide specified. An example is half your parameter descriptions end with a point, the other half doesn't. It's especially useful for class documentation if you're going to use automated doc generation tools later on.  
3) The docs are very light, there is no table of contents and markdown is hard to navigate, the titles are really close in size, hard to parse visually. Some parameters aren't documented at all (below layer for example). I think an automated documentation tool would be really useful and easy to set up.
Documentation was one of my main pain points with Folium, as a daily professional user, I had to regularly plunge into leaflet's JS doc which is really a no go if you want to be efficient.  

&nbsp;


I don't mean to be overly critical, I know the lib is young, but I think those changes could make a big difference in ease of adoption and contribution.

ryanbaumann commented:
This is excellent feedback, @etienneDesticourt.  Thank you!  I added the `pip install mock` and added a style guide (pep 8) to the readme - which you're right, we need to clean up and enforce throughout the library.  

On docs, yes, agreed!  We decided to get the work out about the library with markdown based documentation before diving in with automated Sphinx docs.  That's on our list to hit - I began work on it here https://github.com/mapbox/mapboxgl-jupyter/pull/42.

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -39,18 +39,13 @@ def __init__(self,
         """
         if access_token is None:
             access_token = os.environ.get('MAPBOX_ACCESS_TOKEN', '')
-        if style_url.startswith('mapbox:') and not access_token.startswith('pk'):
```
I'd like to retain this token error to give feedback to users who want to create a map from a Mapbox Style URL but don't specify a valid access token.  

Could we add another check here to see if the type of object passed as a style is a `dict` or a `string`?  If it's a string, we can check if the style is a `mapbox://` format url, and if so, check for a token error.

ryanbaumann commented on tests/test_html.py:
```diff
@@ -16,22 +15,6 @@ def data():
 TOKEN = 'pk.abc123'
 
 
-def test_secret_key_CircleViz(data):
```
Can we add these two tests back in, if we reinstate the access token error for mapbox URL style sheets only (above commnet)?

ryanbaumann commented:
Also noting that we'll want to update all examples to now use `style` instead of `style_url` for all map initialization.

EtienneDesticourt commented:
@ryanbaumann 
Hey Ryan,
I'm working on a proposal to propagate the errors from mapbox.js which would allow us to avoid parsing the style for mapbox-like urls and generally reuse all the exception handling that's already been done there so the maps can't fail silently. 
I'll push by the end of the week so we can discuss it and if it doesn't work for you I'll reinstate the token error.

ryanbaumann commented:
> I'm working on a proposal to propagate the errors from mapbox.js which would allow us to avoid parsing the style for mapbox-like urls and generally reuse all the exception handling that's already been done there so the maps can't fail silently.

That's a brilliant idea @EtienneDesticourt!  Excited to review when you're ready.

perrygeo commented on mapboxgl/viz.py:
```diff
@@ -39,18 +39,13 @@ def __init__(self,
-        if not access_token.startswith('pk'):
-            raise TokenError('Mapbox access token must be public (pk). ' \
-                             'Please sign up at https://www.mapbox.com/signup/ to get a public token. ' \
-                             'If you already have an account, you can retreive your token at https://www.mapbox.com/account/.')
```
I would advise strongly against removing this check. Allowing secret keys to pass through unchecked is a serious security vulnerability. Since notebooks will most likely shared on github etc., we run the significant risk of accidentally exposing a user's credentials.

EtienneDesticourt replied:
In that case I would suggest raising an error if the token starts with "sk", this way the users can still use the maps without a mapbox token, and the js library will take care of error handling if they try to use a mapbox resource, but we still ensure they won't publish secrets.

ryanbaumann replied:
I'm good with throwing an error if the token begins with `sk`.

EtienneDesticourt commented:
I struggled with IPython for a few days to get Javascript to raise a python error using the kernel API but it didn't work out. Instead I propose this compromise where the Javascript replaces the iframe with information (file, stack, message) about the error in the style used by Jupyter notebooks when displaying python exceptions.
It looks like this:
![image](https://user-images.githubusercontent.com/8978248/37237560-14f48ed6-2415-11e8-8a1c-fa9da29f1e33.png)

Also added the token check and reverted the test removal.

ryanbaumann commented on mapboxgl/templates/main.html:
```diff
@@ -126,6 +126,41 @@
+    return formattedError;
+}
+
+window.onerror = function (errorMsg, url, lineNumber, column, errorObj) {
```
this is a common need when the map has a valid style json, but the layer we add for the CircleViz, HeatmapViz, etc is not valid.

Perhaps display the error in a modal the user can dismiss?

Also, we probably only want to show some errors from the map here - not all errors.  For example, the map could throw a warning when requesting an invalid tile from a valid tileset.

EtienneDesticourt replied:
Alright sounds easy enough, I rebased this PR to get rid of the error handling commit so we can close it if you're fine with it.
I'll open a new PR after dinner with the error handling code and modal related modifications.

ryanbaumann replied:
Great!  Thanks @etienneDesticourt.  I think we're ready to merge pending @perrygeo 👍
